tiny_dnn/layers/recurrent_cell_layer.h

```diff
@@ -0,0 +1,169 @@
+             vector_type::aux}) {       // output state
+    set_params(in_dim, out_dim, has_bias);
+    init_backend(backend_type);
+    layer::set_backend_type(backend_type);
+  }
+
+  // move constructor
+  recurrent_cell_layer(recurrent_cell_layer &&other)
+    : layer(std::move(other)),
+      params_(std::move(other.params_)),
+      kernel_fwd_(std::move(other.kernel_fwd_)),
+      kernel_back_(std::move(other.kernel_back_)) {
+    init_backend(std::move(other.engine()));
+  }
+
+  serial_size_t fan_in_size(serial_size_t i) const override {
+    return in_shape()[i].width_;
+  }
+
+  serial_size_t fan_out_size(serial_size_t i) const override {
+    return in_shape()[i].height_;
+  }
+
+  std::vector<index3d<serial_size_t>> in_shape() const override {
+    if (params_.has_bias_) {
+      return {
+        index3d<serial_size_t>(params_.in_size_, 1, 1),   // x
+        index3d<serial_size_t>(params_.out_size_, 1, 1),  // h(t-1)
+        index3d<serial_size_t>(params_.in_size_, params_.out_size_, 1),   // U
+        index3d<serial_size_t>(params_.out_size_, params_.out_size_, 1),  // W
+        index3d<serial_size_t>(params_.out_size_, params_.out_size_, 1),  // V
+        index3d<serial_size_t>(params_.out_size_, 1, 1),                  // b
+        index3d<serial_size_t>(params_.out_size_, 1, 1)};                 // c
+    } else {
+      return {
+        index3d<serial_size_t>(params_.in_size_, 1, 1),   // x
+        index3d<serial_size_t>(params_.out_size_, 1, 1),  // h(t-1)
+        index3d<serial_size_t>(params_.in_size_, params_.out_size_, 1),    // U
+        index3d<serial_size_t>(params_.out_size_, params_.out_size_, 1),   // W
+        index3d<serial_size_t>(params_.out_size_, params_.out_size_, 1)};  // V
+    }
+  }
+
+  std::vector<index3d<serial_size_t>> out_shape() const override {
+    return {index3d<serial_size_t>(params_.out_size_, 1, 1),
+            index3d<serial_size_t>(params_.out_size_, 1, 1)};  // h(t)
+  }
+
+  void forward_propagation(const std::vector<tensor_t *> &in_data,
+                           std::vector<tensor_t *> &out_data) override {
+    // forward recurrent op context
+    auto ctx = OpKernelContext(in_data, out_data);
+    ctx.setParallelize(layer::parallelize());
+    ctx.setEngine(layer::engine());
+
+    // launch recurrent kernel
+    kernel_fwd_->compute(ctx);
+  }
+
+  void back_propagation(const std::vector<tensor_t *> &in_data,
+                        const std::vector<tensor_t *> &out_data,
+                        std::vector<tensor_t *> &out_grad,
+                        std::vector<tensor_t *> &in_grad) override {
+    // backward recurrent op context
+    auto ctx = OpKernelContext(in_data, out_data, out_grad, in_grad);
+    ctx.setParallelize(layer::parallelize());
+    ctx.setEngine(layer::engine());
+
+    // launch recurrent kernel
+    kernel_back_->compute(ctx);
+  }
+
+  std::string layer_type() const override { return "recurrent"; }
+
+#ifndef CNN_NO_SERIALIZATION
```

Comment:
You don't need this preprocessor macro.
Please refer #667 
An empty `struct serialization_buddy` will be there if `CNN_NO_SERIALIZATION` isn't defined so there's no need for `#ifdef CNN_USE_GEMMLOWP` preprocessor directives in layer classes.